sotodlib/site_pipeline/make_focal_plane.py

```diff
@@ -0,0 +1,264 @@
+import sotodlib.io.g3tsmurf_utils as g3u
+from sotodlib.core import AxisManager
+from scipy.spatial.transform import Rotation as R
+from pycpd import AffineRegistration
+
+
+def LAT_coord_transform(xy, rot_fp, rot_ufm, r=72.645):
+    """
+    Transform from instrument model coords to LAT Zemax coords
+
+    Arguments:
+
+        xy: XY coords from instrument model.
+            Should be a (2, n) array.
+
+        rot_fp: Angle of array location on focal plane in deg.
+
+        rot_ufm: Rotatation of UFM about its center.
+
+    Returns:
+
+        xy_trans: Transformed coords.
+    """
+    xy_trans = np.zeros((xy.shape[1], 3))
+    xy_trans[:, :2] = xy.T
+
+    r1 = R.from_euler("z", rot_fp, degrees=True)
+    shift = r1.apply(np.array([r, 0, 0]))
+
+    r2 = R.from_euler("z", rot_ufm, degrees=True)
+    xy_trans = r2.apply(xy_trans) + shift
+
+    return xy_trans.T[:2]
+
+
+def rescale(xy):
+    """
+    Rescale pointing or template to [0, 1]
+
+    Arguments:
+
+        xy: Pointing or template, should have two columns.
+
+    Returns:
+
+        xy_rs: Rescaled array.
+    """
+    xy_rs = xy.copy()
+    xy_rs[:, 0] /= xy[:, 0].max() - xy[:, 0].min()
+    xy_rs[:, 0] -= xy_rs[:, 0].min()
+    xy_rs[:, 1] /= xy[:, 1].max() - xy[:, 1].min()
+    xy_rs[:, 1] -= xy_rs[:, 1].min()
+    return xy_rs
+
+
+def gen_priors(aman, prior, method="flat", width=1, basis=None):
+    """
+    Generate priors from detmap.
+    Currently priors will be 1 everywhere but the location of the detmap's match.
+    More complicated prior generation will be a part of this function eventually.
+
+    Arguments:
+
+        aman: AxisManager assumed to contain aman.det_info.det_id and aman.det_info.wafer.
+
+        prior: Prior value at locations from the detmap.
+               Should be greater than 1.
+
+        method: What sort of priors to implement.
+                Currently only 'flat' is accepted but at least 'gaussian' will be implemented later.
+
+        width: Width of priors. When gaussian priors are added this will be sigma.
+
+        basis: Basis to calculate width in.
+               Currently not implemented so width will just be along the dets axis.
+               At the very least radial distance will be added.
+
+    Returns:
+
+        priors: The 2d array of priors.
+    """
+
+    def _flat(arr, idx):
+        arr[idx - width // 2 : idx + width // 2 + width % 2] = prior
+
+    if method == "flat":
+        prior_method = _flat
+    else:
+        raise ValueError("Method " + method + " not implemented")
+
+    priors = np.ones((aman.dets.count, aman.dets.count))
+    for i in aman.dets.count:
+        prior_method(priors[i], i)
+
+    return priors
+
+
+def match_template(
+    focal_plane, template, out_thresh=0, avoid_collision=True, priors=None
+):
+    """
+    Match fit focal plane againts a template.
+
+    Arguments:
+
+        focal_plane: Focal plane after either having optics model applied or being fit with channel map.
```

Comment:
I assume this focal_plane argument will contain the detmap information?

Reply:
Sorry these argument names haven't changed since a very early version and may be unclear. Docstrings are also out of date, I'll update them soon.
`focal_plane` is the pointing information that you produce.
`template` is the designed `det_x`, `det_y`, and `angle` of the detectors. I currently pull these from the values in the detmap. There are probably other sources that I could pull this from if we want things to work even if there is no detmap provided.